src/uam/io/tcp_client.cpp

```diff
@@ -0,0 +1,251 @@
+  socket_.set_option(option, error_code);
+  if (error_code)
+  {
+    ROS_ERROR_STREAM("Failed to set option: " << error_code.message());
+    disconnect();
+    return false;
+  }
+
+  ROS_INFO_STREAM("Connected to: " << remote_ip << ":" << remote_port);
+
+  int recv_buf_size;
+  socklen_t recv_buf_size_len = sizeof(recv_buf_size);
+  auto rc = getsockopt(socket_.native_handle(), SOL_SOCKET, SO_RCVBUF, &recv_buf_size, &recv_buf_size_len);
+  if (rc == 0)
+  {
+    ROS_INFO_STREAM("Actual receive buffer size: " << recv_buf_size << " bytes");
+  }
+  else
+  {
+    ROS_ERROR_STREAM("Error getting receive buffer size: " << strerror(errno));
+  }
+
+  return true;
+}
+
+void TcpClient::asyncReadData(const size_t packet_offset)
+{
+  if (!connected_ || stop_requested_)
+  {
+    stopped_ = true;
+    return;
+  }
+
+  // Try read  with the minimum reply size
+  boost::asio::async_read(
+    socket_,
+    boost::asio::buffer(
+      receive_buffer_.getRawPacket().data() + packet_offset,
+      sizeof(protocol::CommandReplyHeader) - packet_offset),
+    boost::asio::transfer_exactly(sizeof(protocol::CommandReplyHeader)),
+    boost::bind(
+      &TcpClient::handleReceive,
+      this,
+      boost::asio::placeholders::error,
+      boost::asio::placeholders::bytes_transferred));
+}
+
+void TcpClient::handleReceive(const boost::system::error_code& error_code, size_t bytes_transferred)
+{
+  // Grab a packet received time as soon as possible
+  auto received_stamp = ros::Time::now();
+  size_t packet_offset = 0;
+
+  if (error_code)
+  {
+    // "operation_aborted" errors will occur if the socket is closed for any reason. Don't spam the logs with expected
+    // error conditions.
+    if (connected_ || error_code != boost::asio::error::operation_aborted)
+    {
+      ROS_ERROR_STREAM("Error receiving lidar message. " << error_code.message());
+    }
+  }
+  else if (sizeof(protocol::CommandReplyHeader) != bytes_transferred)
+  {
+    ROS_WARN_STREAM(
+      "Failed to read: " << error_code.message() << " Received " << bytes_transferred << "bytes\n"
+                         << "Expected " << sizeof(protocol::CommandReplyHeader) << " bytes");
+  }
+  else
+  {
+    // All the packets are expected to contain a header. We read those bytes
+    // and then we call the read to read the rest
+    bool should_try_complete_read =
+      filter_callback_ ? filter_callback_(receive_buffer_.get<protocol::CommandReplyHeader>().header) : true;
+
+    if (should_try_complete_read)
+    {
+      auto stamp = ros::Time::now();
+      auto expected_total_size = receive_buffer_.get<protocol::CommandReplyHeader>().cmd_size;
+      auto valid_expected_size =
+        decodeField(expected_total_size) && expected_total_size <= sizeof(protocol::AR01CommandReply);
+
+      if (valid_expected_size)
+      {
+        boost::system::error_code new_error_code;
+        auto missing_read = expected_total_size - bytes_transferred;
+        auto recv_bytes = boost::asio::read(
```

Comment:
Is this read() okay? Even though you are using async writes elsewhere?

Reply:
Here I am not sure. I need to read the missing bytes. It was so much easier to just block the worker thread right away. But I am not sure if I am violating any of the boost::asio principles...